File: 200. Number of Islands.cpp
class Solution {
    int m, n;
    vector<pair<int, int>> dirs = {{0, 1}, {0, -1}, {-1, 0}, {1, 0}};
    
    void dfs(int x, int y, vector<vector<char>> &grid) {
        grid[x][y] = '0';
        
        for(auto &dir : dirs) {
            int dx = x + dir.first, dy = y + dir.second;
            
            if(dx >= 0 and dx < m and dy >= 0 and dy < n and grid[dx][dy] == '1')
                dfs(dx, dy, grid);
        }
    }
public:
    int numIslands(vector<vector<char>>& grid) {
        m = grid.size(), n = grid[0].size();
        int ans = 0;
        
        for(int i = 0; i < m; i++) {
            for(int j = 0; j < n; j++) {
                if(grid[i][j] == '1') {
                    dfs(i, j, grid);
                    ans++;
                }
            }
        }
        
        return ans;
    }
};

File: 200. Number of Islands.py
class Solution:
    def dfs(self, grid, m, n, i, j):
        grid[i][j] = "0"

        if(i + 1 < m and grid[i + 1][j] == "1"):
            self.dfs(grid, m, n, i + 1, j)
        if(i - 1 >= 0 and grid[i - 1][j] == "1"):
            self.dfs(grid, m, n, i - 1, j)
        if(j + 1 < n and grid[i][j + 1] == "1"):
            self.dfs(grid, m, n, i, j + 1)
        if(j - 1 >= 0 and grid[i][j - 1] == "1"):
            self.dfs(grid, m, n, i, j - 1)


    def numIslands(self, grid: List[List[str]]) -> int:
        m, n = len(grid), len(grid[0])
        ans = 0

        for i in range(m):
            for j in range(n):
                if(grid[i][j] == "1"):
                    ans += 1
                    self.dfs(grid, m, n, i, j)
        
        return ans
        
